Track and manage Opaque Data Types in CoreIPC

sheeparegreat opened the pull request:
#### b21c169a1e5eac2379544b133ea2ee7a038c979a
<pre>
Track and manage Opaque Data Types in CoreIPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=299055">https://bugs.webkit.org/show_bug.cgi?id=299055</a>
<a href="https://rdar.apple.com/160811093">rdar://160811093</a>

Reviewed by Alex Christensen and Mike Wyrzykowski.

The goal of this change is to track and manage Opaque Data Types (which represent a sandbox escape risk) in the CoreIPC flows.

Tracking of Opaque Data Types is done via Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in

What is considered an Opaque Data Type can be seen in Source/WebKit/Scripts/webkit/opaque_ipc_types.py.

Test: Source/WebKit/Scripts/webkit/opaque_ipc_types.py
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-message-receiver.py:
(main):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.enforce_opaque_ipc_types_usage):
(UsingStatement.enforce_opaque_ipc_types_usage):
(encode_type):
(main):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.enforce_opaque_ipc_types_usage):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.py: Added.
(_is_odt_concern):
(_remove_const_and_whitespace):
(_split_template_parameters):
(_array_special_parsing):
(_get_container_info):
(_get_parameters_to_check):
(_contains_opaque_data):
(OpaqueIPCTypeEntry):
(OpaqueIPCTypeEntry.__init__):
(OpaqueIPCTypeEntry._parse_attributes):
(OpaqueIPCTypes):
(OpaqueIPCTypes.__init__):
(OpaqueIPCTypes._parse_group_header):
(OpaqueIPCTypes._parse_alias_param):
(OpaqueIPCTypes._parse_structure_param):
(OpaqueIPCTypes._parse_message_param):
(OpaqueIPCTypes._parse_line):
(OpaqueIPCTypes._add_entry):
(OpaqueIPCTypes._query_entries):
(OpaqueIPCTypes._is_webcontent_dispatchable):
(OpaqueIPCTypes.message_param_tracked):
(OpaqueIPCTypes.message_param_reply_tracked):
(OpaqueIPCTypes.alias_param_tracked):
(OpaqueIPCTypes.structure_param_tracked):
(OpaqueIPCTypes.webcontent_dispatchable):
(OpaqueIPCTypes.reply_webcontent_dispatchable):
(OpaqueIPCTypes.structure_webcontent_dispatchable):
(is_opaque_type):
(TestOpaqueTypes):
(TestOpaqueTypes.test_is_odt_concern_function):
(TestOpaqueTypes.test_contains_opaque_data_function):
(TestOpaqueTypes.test_direct_opaque_types):
(TestOpaqueTypes.test_container_types_with_odt_concerns):
(TestOpaqueTypes.test_container_types_without_odt_concerns):
(TestOpaqueTypes.test_infinite_recursion_protection):
(TestOpaqueTypes.test_bad_formatting):
(TestOpaqueTypes.test_context_propagation_through_simple_wrappers):
(TestOpaqueTypes.test_context_reset_in_structural_containers):
(TestOpaqueTypes.test_retainptr_with_direct_opaque_types):
(TestOpaqueTypes.test_deeply_nested_types):
(TestOpaqueTypes.test_opaque_ipc_types_parsing):
(TestOpaqueTypes.test_production_tracking_file_parses):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in: Added.
* Source/WebKit/Scripts/webkit/parser_unittest.py:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::OtherClass&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::OpaqueTypeObject&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::OpaqueTypeObject&gt;::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiver.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(WebKit::TestWithLegacyReceiver::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertion&gt;):
(IPC::jsValueForDecodedMessageReply&lt;MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertion&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertionReply&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::name):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::asyncMessageReplyName):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::OpaqueTypeSecurityAssertion):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::encode):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::name):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::OpaqueTypeSecurityAssertionReply):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::encode):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributes.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(WebKit::TestWithoutAttributes::didReceiveMessage):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertion&gt;):
(IPC::jsValueForDecodedMessageReply&lt;MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertion&gt;):
(IPC::jsValueForDecodedMessage&lt;MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertionReply&gt;):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::name):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::asyncMessageReplyName):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::OpaqueTypeSecurityAssertion):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::encode):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::name):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::OpaqueTypeSecurityAssertionReply):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::encode):
* Source/WebKit/Scripts/webkit/tests/test_opaque_ipc_types.tracking.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/303390@main">https://commits.webkit.org/303390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76df893b2298655f9e2a480059dbab0d344ad0ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4618 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43147 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139640 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84041 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8a1c928c-62b4-4d47-bc95-f99c4f9f6f31) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4728 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4379 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101000 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d2a491a7-3050-4bb1-9bac-35d84c0e8ad5) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135071 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118345 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81791 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0587736b-3619-47b4-a463-c0ee1d9192dc) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/131474 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82859 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124188 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/112046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142286 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130632 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4287 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37044 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109374 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4368 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3731 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109549 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27774 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3256 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114620 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57533 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4341 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33013 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163599 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4172 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67787 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42518 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4432 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4300 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sheeparegreat commented on Source/WebKit/Scripts/webkit/opaque_ipc_types.py:
```diff
@@ -0,0 +1,598 @@
+            return True
+        return any(e.type == type_string for e in self.structure_params[key])
+
+    def webcontent_dispatchable(self, receiver, message, parameter_name, parameter_type):
```
Update method name to message_param_webcontent_dispatchable

sheeparegreat commented on Source/WebKit/Shared/OpaqueTransports.tracking.in:
```diff
@@ -0,0 +1,277 @@
+[Legacy] MessageParamReply RemoteGraphicsContextGL.GetBufferSubDataInline data std::span<const uint8_t>
+[Legacy] MessageParamReply RemoteGraphicsContextGL.PrepareForDisplay displayBuffer MachSendRight
+[Legacy] MessageParamReply RemoteGraphicsContextGL.ReadPixelsInline data std::span<const uint8_t>
+[ProvenUnstructuredData='Pixel buffer data'] MessageParam RemoteGraphicsContextGL.TexImage2D0 pixels std::span<const uint8_t>
```
Label these as Legacy

sheeparegreat commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -465,6 +477,22 @@ def __init__(self, name, alias_lines, condition):
 
+    def enforce_opaque_transports_usage(self):
+        if opaque_transports is None:
+            return  # Tracking file not available, skip enforcement
```
Throw here

sheeparegreat commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -832,6 +860,8 @@ def encode_type(type):
             result.append(f'        encoder << instance.{member.name};')
         else:
+            if not opaque_transports.structure_webcontent_dispatchable(type.namespace_and_name(), member.name, member.type):
+                result.append('    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isInWebProcess());')
```
ASSERT instead of RELEASE_ASSERT

sheeparegreat commented on Source/WebKit/Shared/OpaqueTransports.tracking.in:
```diff
@@ -0,0 +1,476 @@
+    [Legacy] MessageParamReply TestWithStream.SendAndReceiveMachSendRight r1
+}
+[NotSentFromWebContent] StructureParam WebCore::OpaqueTypeObject.member NotDispatchableFromWebContent
+[NotDispatchableFromWebContent] MessageParam TestWithLegacyReceiver.OpaqueTypeSecurityAssertion ping NotDispatchableFromWebContent
```
NotSentFromWebContent

sheeparegreat commented on Source/WebKit/Scripts/webkit/opaque_types.py:
```diff
@@ -0,0 +1,735 @@
+                    elif key == 'Docs' or key == 'docs':
+                        self.docs = value.strip("'\"")
+                else:
+                    if attribute == 'NotDispatchableFromWebContent':
```
Remove this old name

sheeparegreat commented on Source/WebKit/Shared/OpaqueTransports.tracking.in:
```diff
@@ -0,0 +1,476 @@
+    [NotSentFromWebContent] MessageParam WebSocketChannel.DidReceiveBinaryData data
+    [NotSentFromWebContent] MessageParam RTCDataChannelRemoteManager.ReceiveData data
+    [NotSentFromWebContent] MessageParam RTCDataChannelRemoteManager.SendData text
+    [NotSentFromWebContent] MessageParam RTCDataChannelRemoteManagerProxy.ReceiveData data
```
`ASSERTION FAILED: !isInWebProcess()
/Volumes/Data/worker/macOS-Tahoe-Debug-Build-EWS/build/WebKitBuild/Debug/DerivedSources/WebKit/RTCDataChannelRemoteManagerProxyMessages.h(172) : void Messages::RTCDataChannelRemoteManagerProxy::ReceiveData::encode(Encoder &) [Encoder = IPC::Encoder]
1   0x11b318628 void Messages::RTCDataChannelRemoteManagerProxy::ReceiveData::encode<IPC::Encoder>(IPC::Encoder&)
2   0x11b2fc924 IPC::Error IPC::Connection::send<Messages::RTCDataChannelRemoteManagerProxy::ReceiveData>(Messages::RTCDataChannelRemoteManagerProxy::ReceiveData&&, unsigned long long, WTF::OptionSet<IPC::SendOption, (WTF::ConcurrencyTag)0>, std::__1::optional<WTF::Thread::QOS>)
3   0x11b2fc8a0 WebKit::RTCDataChannelRemoteManager::RemoteSourceConnection::didReceiveStringData(WebCore::ProcessQualified<WTF::ObjectIdenti`

sheeparegreat replied:
Test: http/wpt/webrtc/transfer-datachannel-service-worker.https.html

sheeparegreat commented on Source/WebKit/DerivedSources-input.xcfilelist:
```diff
@@ -207,9 +207,12 @@ $(PROJECT_DIR)/Scripts/generate-derived-log-sources.py
 $(PROJECT_DIR)/Scripts/generate-https-upgrade-database.sh
 $(PROJECT_DIR)/Scripts/generate-message-receiver.py
 $(PROJECT_DIR)/Scripts/generate-serializers.py
+$(PROJECT_DIR)/Scripts/webkit/OpaqueTransports.tracking.in
```
Remove this

achristensen07 commented on Source/WebKit/Scripts/webkit/opaque_transports.py:
```diff
@@ -0,0 +1,804 @@
+    "MachSendRight", "MachSendRightAnnotated", "WTF::MachSendRightAnnotated",
+    # NSArray to be added in a follow-up change
+    "WebCore::SharedMemoryHandle",
+    "WebCore::SharedMemory::Handle",
```
This has been removed from IPC.  I think they're all WebCore::SharedMemoryHandle now.

mwyrzykowski replied:
Aren't there still a few `SharedMemory::Handle` usages?

https://github.com/WebKit/WebKit/blob/5637407989f35220eb774158c20490e0a3083904/Source/WebKit/Shared/FullScreenMediaDetails.serialization.in#L35
https://github.com/WebKit/WebKit/blob/5637407989f35220eb774158c20490e0a3083904/Source/WebKit/GPUProcess/graphics/WebGPU/RemoteQueue.messages.in#L36
https://github.com/WebKit/WebKit/blob/5637407989f35220eb774158c20490e0a3083904/Source/WebKit/GPUProcess/graphics/WebGPU/RemoteBuffer.messages.in#L34
https://github.com/WebKit/WebKit/blob/5637407989f35220eb774158c20490e0a3083904/Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.messages.in#L35

achristensen07 commented on Source/WebKit/Scripts/webkit/opaque_ipc_types.py:
```diff
@@ -0,0 +1,804 @@
+    "CFDataRef", "NSData", "WebKit::CoreIPCData",
+    "CFDictionaryRef", "CFArrayRef", "NSDictionary",
+    "WebCore::SharedMemoryHandle",
+    "WebCore::SharedMemory::Handle",
```
This is a redundant list.  Can one list add the other like we do in types_that_cannot_be_forward_declared?

achristensen07 commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -25,6 +25,7 @@
 import re
 import sys
 
+from webkit.opaque_transports import opaque_transports
```
I don't really like the use of the word "transports" in this PR.  Can it be opaque_ipc_types?

sheeparegreat committed:
Track and manage Opaque Data Types in CoreIPC
https://bugs.webkit.org/show_bug.cgi?id=299055
rdar://160811093

Reviewed by Alex Christensen and Mike Wyrzykowski.

The goal of this change is to track and manage Opaque Data Types (which represent a sandbox escape risk) in the CoreIPC flows.

Tracking of Opaque Data Types is done via Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in

What is considered an Opaque Data Type can be seen in Source/WebKit/Scripts/webkit/opaque_ipc_types.py.

Test: Source/WebKit/Scripts/webkit/opaque_ipc_types.py
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-message-receiver.py:
(main):
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.enforce_opaque_ipc_types_usage):
(UsingStatement.enforce_opaque_ipc_types_usage):
(encode_type):
(main):
* Source/WebKit/Scripts/webkit/messages.py:
(message_to_struct_declaration):
* Source/WebKit/Scripts/webkit/model.py:
(MessageReceiver.enforce_opaque_ipc_types_usage):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.py: Added.
(_is_odt_concern):
(_remove_const_and_whitespace):
(_split_template_parameters):
(_array_special_parsing):
(_get_container_info):
(_get_parameters_to_check):
(_contains_opaque_data):
(OpaqueIPCTypeEntry):
(OpaqueIPCTypeEntry.__init__):
(OpaqueIPCTypeEntry._parse_attributes):
(OpaqueIPCTypes):
(OpaqueIPCTypes.__init__):
(OpaqueIPCTypes._parse_group_header):
(OpaqueIPCTypes._parse_alias_param):
(OpaqueIPCTypes._parse_structure_param):
(OpaqueIPCTypes._parse_message_param):
(OpaqueIPCTypes._parse_line):
(OpaqueIPCTypes._add_entry):
(OpaqueIPCTypes._query_entries):
(OpaqueIPCTypes._is_webcontent_dispatchable):
(OpaqueIPCTypes.message_param_tracked):
(OpaqueIPCTypes.message_param_reply_tracked):
(OpaqueIPCTypes.alias_param_tracked):
(OpaqueIPCTypes.structure_param_tracked):
(OpaqueIPCTypes.webcontent_dispatchable):
(OpaqueIPCTypes.reply_webcontent_dispatchable):
(OpaqueIPCTypes.structure_webcontent_dispatchable):
(is_opaque_type):
(TestOpaqueTypes):
(TestOpaqueTypes.test_is_odt_concern_function):
(TestOpaqueTypes.test_contains_opaque_data_function):
(TestOpaqueTypes.test_direct_opaque_types):
(TestOpaqueTypes.test_container_types_with_odt_concerns):
(TestOpaqueTypes.test_container_types_without_odt_concerns):
(TestOpaqueTypes.test_infinite_recursion_protection):
(TestOpaqueTypes.test_bad_formatting):
(TestOpaqueTypes.test_context_propagation_through_simple_wrappers):
(TestOpaqueTypes.test_context_reset_in_structural_containers):
(TestOpaqueTypes.test_retainptr_with_direct_opaque_types):
(TestOpaqueTypes.test_deeply_nested_types):
(TestOpaqueTypes.test_opaque_ipc_types_parsing):
(TestOpaqueTypes.test_production_tracking_file_parses):
* Source/WebKit/Scripts/webkit/opaque_ipc_types.tracking.in: Added.
* Source/WebKit/Scripts/webkit/parser_unittest.py:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::OtherClass>::encode):
(IPC::ArgumentCoder<WebCore::OpaqueTypeObject>::encode):
(IPC::ArgumentCoder<WebCore::OpaqueTypeObject>::decode):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/Makefile:
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::jsValueForArguments):
(IPC::jsValueForReplyArguments):
(IPC::messageArgumentDescriptions):
(IPC::messageReplyArgumentDescriptions):
* Source/WebKit/Scripts/webkit/tests/MessageNames.cpp:
* Source/WebKit/Scripts/webkit/tests/MessageNames.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiver.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(WebKit::TestWithLegacyReceiver::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertion>):
(IPC::jsValueForDecodedMessageReply<MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertion>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithLegacyReceiver_OpaqueTypeSecurityAssertionReply>):
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::name):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::asyncMessageReplyName):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::OpaqueTypeSecurityAssertion):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertion::encode):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::name):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::OpaqueTypeSecurityAssertionReply):
(Messages::TestWithLegacyReceiver::OpaqueTypeSecurityAssertionReply::encode):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributes.messages.in:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(WebKit::TestWithoutAttributes::didReceiveMessage):
(IPC::jsValueForDecodedMessage<MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertion>):
(IPC::jsValueForDecodedMessageReply<MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertion>):
(IPC::jsValueForDecodedMessage<MessageName::TestWithoutAttributes_OpaqueTypeSecurityAssertionReply>):
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::name):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::asyncMessageReplyName):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::OpaqueTypeSecurityAssertion):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertion::encode):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::name):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::OpaqueTypeSecurityAssertionReply):
(Messages::TestWithoutAttributes::OpaqueTypeSecurityAssertionReply::encode):
* Source/WebKit/Scripts/webkit/tests/test_opaque_ipc_types.tracking.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/303390@main
